raspberry/checkout/checkout_main.py

```diff
@@ -62,16 +65,40 @@ def process_rfid_card(self, uid):
             print("Item already scanned")
             self.interactions.buzz()
             self.interactions.indicate_error()
+
+    def cancel_action(self):
+        if self.state == 1:
+            return self.cancel_checkout
+        else:
+            return self.logic.remove_current_tag
+
+    def confirm_action(self):
+        if self.state == 1:
+            return self.finish_checkout
+        else:
+            def change_state():
```

Comment:
this is way over-engineered, you create a function inside a function and then you return this function without ever executing it. all you need to do is set self.state to 1. the function in the current scenario is never called.

Reply:
on, and in the finish checkout function the state must be reset to 0. after all after finishing shopping, we want to be able to shop again. same goes for the cancel_checkout function. after all, if buying is not successful on the server-side and we cancel the checkout, which is mentioned in one of the previous comments, we should also call the self.cancel_checkout function

Reply:
oh, an after changing the state, you should probably display info something like: SUMMARY - use encoder to see added products, red btn to remove, green to confirm order.
so that the user gets feedback that pressing green actually did anything